x/concentrated-liquidity/swaps.go

```diff
@@ -538,29 +549,28 @@ func (k Keeper) computeInAmtGivenOut(
 
 		// update the swapState with the new sqrtPrice from the above swap
 		swapState.sqrtPrice = sqrtPrice
-		swapState.amountSpecifiedRemaining = swapState.amountSpecifiedRemaining.Sub(amountOut)
-		swapState.amountCalculated = swapState.amountCalculated.Add(amountIn.Add(feeChargeTotal))
+		swapState.amountSpecifiedRemaining = swapState.amountSpecifiedRemaining.SubMut(amountOut)
```

Comment:
what is the reason using mut here improves performance?

Reply:
Basically, every Int/Decimal operation in the SDK allocates a new big.Int

Every new big.Int requires a heap allocation, which takes more time. (And then has to get garbage collected later)

This page talks a bit about perf differences stack/heap: https://stackoverflow.com/questions/36512509/why-is-stack-memory-cheaper-than-heap-memory-in-c  (and keep in mind that in C you manually free things. Garbage collection is far more expensive than a free call)

Lmk if your unfamiliar with stack/heap! (In which case the above post won't make any sense)

Reply:
So this line is going from `a = a - b` to instead `a -= b`. Now in every single compiled language, these will have no performance differences for normal code.

However in golang, we don't have interfaces for basic arithmetic that can get compiler optimized. So it can't infer that `a = a.Add(b)` is equivalent to `a += b`, and the internal methods are slower.

Rust doesn't have this problem

Reply:
Ah https://github.com/osmosis-labs/osmosis/pull/5211#discussion_r1197498713 this change was the question that I had in mind, everythings clear now thanks!